apps/judicial-system/api/src/app/modules/auth/auth.controller.ts

```diff
@@ -215,138 +239,219 @@ export class AuthController {
+        eligibleUsers,
+        userIdx,
+        res,
+        csrfToken === 'undefined' ? undefined : csrfToken,
+      )
+    } catch (error) {
+      this.logger.error('Login failed:', { error })
+
+      this.clearCookies(res)
+
+      res.redirect('/?villa=innskraning-villa')
     }
+  }
 
-    if (authorization) {
-      this.backendService.createEventLog({
-        eventType: loginBypass ? EventType.LOGIN_BYPASS : EventType.LOGIN,
-        nationalId: user?.nationalId,
-        userRole: user?.role,
-        userName: user?.name,
-        userTitle: user?.title,
-        institutionName: user?.institution?.name,
-      })
+  private clearCookies(
+    res: Response,
+    cookies: Cookie[] = [
+      this.codeVerifierCookie,
+      this.idToken,
+      this.redirectCookie,
+      this.csrfCookie,
+      this.accessTokenCookie,
+    ],
+  ) {
+    const clearCookie = (cookie: Cookie) => {
+      res.req.cookies[cookie.name] &&
+        res.clearCookie(cookie.name, cookie.options)
+    }
 
-      return authorization
+    cookies.forEach((cookie) => clearCookie(cookie))
+  }
+
+  private async findEligibleUsersByNationalId(nationalId: string) {
+    const users = await this.authService.findUsersByNationalId(nationalId)
+
+    if (users) {
+      return users
     }
 
-    this.backendService.createEventLog({
-      eventType: loginBypass
-        ? EventType.LOGIN_BYPASS_UNAUTHORIZED
-        : EventType.LOGIN_UNAUTHORIZED,
-      nationalId: authUser.nationalId,
-    })
+    const defender = await this.authService.findDefender(nationalId)
 
-    return undefined
+    if (defender) {
+      return [defender]
+    }
+
+    return []
   }
 
   private async redirectAuthenticatedUser(
-    authUser: AuthUser,
+    nationalId: string,
     res: Response,
-    requestedRedirectRoute: string,
+    requestedRedirectRoute?: string,
     idToken?: string,
     csrfToken?: string,
   ) {
-    const authorization = await this.authorizeUser(
-      authUser,
+    const eligibleUsers = await this.findEligibleUsersByNationalId(nationalId)
+
+    if (eligibleUsers.length === 0) {
+      this.logger.info('Blocking login attempt from an unauthorized user')
+
+      this.clearCookies(res)
+
+      res.redirect('/?villa=innskraning-ekki-notandi')
+
+      return
+    }
+
+    const currentUserIdx = eligibleUsers.length === 1 ? 0 : -1
+
+    if (idToken) {
+      res.cookie(this.idToken.name, idToken, {
+        ...this.idToken.options,
+        maxAge: EXPIRES_IN_MILLISECONDS,
+      })
+    } else {
+      this.clearCookies(res, [this.idToken])
+    }
+
+    this.redirectAuthorizeUser(
+      eligibleUsers,
+      currentUserIdx,
+      res,
       csrfToken,
       requestedRedirectRoute,
-      !idToken,
     )
+  }
 
-    if (!authorization) {
-      this.logger.info('Blocking login attempt from an unauthorized user')
+  private redirectAuthorizeUser(
+    eligibleUsers: User[], // eligibleUsers is an array of one or more users
```

Comment:
Did we consider storing the users as a different data structure? For example using a `map` where the user Id would be the key? Then we wouldn´t have to worry about the array index and could perform a more efficient lookup.

Reply:
I didn't consider that and I don't think it is necessary. The number of eligible users will be 1 most of the time and probably never more than two.

I did consider storing the current user rather than the current index and may make that change in future PRs.